src/index/blockfilter.cpp

```diff
@@ -0,0 +1,289 @@
+{
+    CBlockUndo block_undo;
+    uint256 prev_header;
+
+    if (pindex->nHeight > 0) {
+        if (!UndoReadFromDisk(block_undo, pindex)) {
+            return false;
+        }
+
+        std::pair<uint256, uint256> read_out;
+        std::pair<char, int> prev_height_key(DB_BLOCK_HEIGHT, pindex->nHeight - 1);
+        if (!m_db->Read(std::make_pair(DB_FILTER_HEADER, prev_height_key), read_out)) {
+            return false;
+        }
+
+        uint256 expected_block_hash = pindex->pprev->GetBlockHash();
+        if (read_out.first != expected_block_hash) {
+            return error("%s: previous block header belongs to unexpected block %s; expected %s",
+                         __func__, read_out.first.ToString(), expected_block_hash.ToString());
+        }
+
+        prev_header = read_out.second;
+    }
+
+    BlockFilter filter(m_filter_type, block, block_undo);
+    std::pair<char, int> height_key(DB_BLOCK_HEIGHT, pindex->nHeight);
+
+    CDBBatch batch(*m_db);
+    batch.Write(std::make_pair(DB_FILTER, height_key),
+                std::make_pair(pindex->GetBlockHash(), filter.GetEncodedFilter()));
+    batch.Write(std::make_pair(DB_FILTER_HASH, height_key),
+                std::make_pair(pindex->GetBlockHash(), filter.GetHash()));
+    batch.Write(std::make_pair(DB_FILTER_HEADER, height_key),
+                std::make_pair(pindex->GetBlockHash(), filter.ComputeHeader(prev_header)));
+    return m_db->WriteBatch(batch);
+}
+
+template <typename T>
+static bool CopyHeightIndexToHashIndex(CDBIterator& db_it, CDBBatch& batch,
+                                       const std::string& index_name, char key_prefix,
+                                       int start_height, const CBlockIndex* stop_index)
+{
+    db_it.Seek(std::make_pair(key_prefix, std::make_pair(DB_BLOCK_HEIGHT, start_height)));
+
+    for (int height = start_height; height <= stop_index->nHeight; ++height) {
+        std::pair<char, std::pair<char, int>> key, expected_key;
+        expected_key = std::make_pair(key_prefix, std::make_pair(DB_BLOCK_HEIGHT, height));
+
+        if (!db_it.GetKey(key) || key != expected_key) {
+            return error("%s: unexpected key in %s: expected (%c, (%c, %d))",
+                         __func__, index_name, key_prefix, DB_BLOCK_HEIGHT, height);
+        }
+
+        std::pair<uint256, T> value;
+        if (!db_it.GetValue(value)) {
+            return error("%s: unable to read value in %s at key (%c, (%c, %d))",
+                         __func__, index_name, key_prefix, DB_BLOCK_HEIGHT, height);
+        }
+
+        auto hash_key = std::make_pair(key_prefix, std::make_pair(DB_BLOCK_HASH, value.first));
+        batch.Write(hash_key, std::move(value.second));
+
+        db_it.Next();
+    }
+    return true;
+}
+
+bool BlockFilterIndex::Rewind(const CBlockIndex* current_tip, const CBlockIndex* new_tip)
+{
+    assert(current_tip->GetAncestor(new_tip->nHeight) == new_tip);
```

Comment:
Assertions should not have side effects. Please move `GetAncestor` outside of assertion :-)

Reply:
What is the side effect? `GetAncestor` is a `const` method.

Reply:
@jimpo You're right! Forget my comment :-)